src/darwin/Framework/CHIP/MTREndpointInfo.mm

```diff
@@ -0,0 +1,211 @@
+            if (endpointID == root->_endpointID) {
+                break; // visited the root, DFS traversal done
+            }
+        } else /* endpoint->_mark == EndpointMark::Visited */ {
+            // Nodes can be visited multiple times due to Full-Family
+            // ancestors like the root node, or in scenarios where a
+            // nodes is erroneously in the PartsList of two separate
+            // branches of the tree. There is no easy way to distinguish
+            // these cases here, so we are not setting valid = NO.
+            deque.pop_front(); // nothing else to do
+        }
+    }
+    if (deque.size() != endpoints.count) {
+        MTR_LOG_ERROR("Warning: Not all endpoints are descendants of the root endpoint");
+        valid = NO;
+    }
+
+    // Now iterate over the endpoints in reverse topological order, i.e. bottom up. This means
+    // that we will visit children before parents, so the first time we see an endpoint in a
+    // PartsList we can assign it as a child of the endpoint we're processing, and we can be sure
+    // that this is the closest parent, not some higher ancestor using the Full-Family Pattern.
+    NSMutableArray<MTREndpointInfo *> * children = [[NSMutableArray alloc] init];
+    while (!deque.empty()) {
+        EndpointId endpointID = deque.front();
+        MTREndpointInfo * endpoint = endpoints[@(endpointID)];
+        deque.pop_front();
+
+        if (endpoint->_mark == EndpointMark::ParentAssigned) {
+            continue; // This endpoint is part of a cycle, don't populate its children.
+        }
+
+        [children removeAllObjects];
+        for (NSNumber * partNumber in endpoint->_partsList) {
+            MTREndpointInfo * descendant = endpoints[partNumber];
+            if (descendant != nil && descendant->_mark != EndpointMark::ParentAssigned) {
+                descendant->_mark = EndpointMark::ParentAssigned;
+                [children addObject:descendant];
+            }
+        }
+        endpoint->_children = [children copy];
+    }
+    root->_mark = EndpointMark::ParentAssigned;
+    return valid;
+}
+
++ (NSDictionary<NSNumber *, MTREndpointInfo *> *)endpointsFromAttributeCache:(const ClusterStateCache *)cache
+{
+    VerifyOrReturnValue(cache != nullptr, nil);
+    using namespace Descriptor::Attributes;
+
+    NSMutableDictionary<NSNumber *, MTREndpointInfo *> * endpoints = [[NSMutableDictionary alloc] init];
+    cache->ForEachAttribute(Descriptor::Id, [&](const ConcreteAttributePath & path) -> CHIP_ERROR {
+        VerifyOrReturnError(path.mAttributeId == DeviceTypeList::Id, CHIP_NO_ERROR);
+
+        CHIP_ERROR err = CHIP_NO_ERROR;
+        NSArray<MTRDescriptorClusterDeviceTypeStruct *> * deviceTypeList = MTRDecodeAttributeValue(path, *cache, &err);
+        if (!deviceTypeList) {
+            MTR_LOG_ERROR("Ignoring invalid DeviceTypeList for endpoint %u: %" CHIP_ERROR_FORMAT, path.mEndpointId, err.Format());
+        }
```

Comment:
Document that leaving deviceTypeList nil is OK, since it will just mean the loop below has no iterations?